### PR TITLE
Forensics

### PR DESCRIPTION
**Pull request by ike709**
## About the PR 

Ports forensics from Nyanotrasen https://github.com/Nyanotrasen/Nyanotrasen/pull/29 which is a remake of https://github.com/space-wizards/space-station-14/pull/7346

Whenever people interact with things, they will leave behind either fingerprints or glove fibers. The forensic scanner will show any prints/fibers on something, and the forensic pad can be used to acquire a person's fingerprints for comparison.

I additionally added a print button to the forensic scanner to print a piece of paper with its current data, a la SS13.

The Nyano PR added the detective job at the same time, but I'm punting that bikeshedding to a future PR.

Credit to @Elijahrane for a lot of the work and @Soupstraw for the original PR.

**Screenshots**

**Changelog**

:cl: Rane, Soupstraw, and ike709
- add: Added a forensics system. You now leave behind fingerprints or glove fibers when you interact with things.
- add: Added the forensic scanner, a tool for detecting prints/fibers on things.
- add: Added forensic pads, used for fingerprinting a suspect.

### CODE REVIEW

**Comment by ike709** [comment]
Conflicts fixed

File: Resources/Locale/en-US/forensics/forensics.ftl [hunk 1]
```diff
@@ -0,0 +1,15 @@
+forensic-scanner-interface-title = Forensic scanner
+forensic-scanner-interface-fingerprints = Fingerprints
+forensic-scanner-interface-fibers = Fibers
+forensic-scanner-interface-no-data = No scan data available
+forensic-scanner-interface-print = Print
+forensic-scanner-report-title = Forensics Report
+forensic-pad-unused = It hasn't been used.
+forensic-pad-sample = It has a sample: {$sample}
+forensic-pad-gloves = Ask {CAPITALIZE($target)} to take off {POSS-ADJ($target)} gloves.
```
Comment (Veritius) [inline Resources/Locale/en-US/forensics/forensics.ftl:9]: Probably could say "target is wearing gloves" as you may not want to bother asking them in certain situations.

Comment (ike709) [reply Resources/Locale/en-US/forensics/forensics.ftl:9]:
```suggestion
forensic-pad-gloves = {CAPITALIZE($target)} is wearing gloves.
```

**Comment by ike709** [comment]
Conflicts fixed

File: Resources/Prototypes/Entities/Objects/Devices/forensic_scanner.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+- type: entity
+  name: forensic scanner
+  parent: BaseItem
+  id: ForensicScanner
+  description: A handheld device that can scan objects for fingerprints and fibers.
+  components:
+  - type: Sprite
+    sprite: Objects/Devices/forensic_scanner.rsi
+    state: forensicnew
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Objects/Devices/forensic_scanner.yml:9]: netsync false

File: Resources/Locale/en-US/forensics/fibers.ftl [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+fibers-leather-green = green leather fibers
+fibers-leather-yellow = yellow leather fibers
+fibers-leather-brown = brown leather fibers
+fibers-leather-black = black leather fibers
+fibers-durathread-regal-blue = regal blue durathread fibers
+fibers-durathread = durathread fibers
+fibers-latex = latex fibers
+fibers-nitrile = nitrile fibers
+fibers-insulative-nanomachines = insulative fibers and nanomachines, son
+fibers-insulative-black = insulative black fibers
```
Comment (metalgearsloth) [inline Resources/Locale/en-US/forensics/fibers.ftl:24]: Just pass in the colour as args?

File: Content.Server/Forensics/Components/ForensicScannerComponent.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+{
+    [RegisterComponent]
+    public sealed class ForensicScannerComponent : Component
+    {
+        public CancellationTokenSource? CancelToken;
+
+        [ViewVariables(VVAccess.ReadOnly)]
+        public List<string> Fingerprints = new();
+        [ViewVariables(VVAccess.ReadOnly)]
+        public List<string> Fibers = new();
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Components/ForensicScannerComponent.cs:13]: Comments?

**Review by mirrorcult — CHANGES_REQUESTED** [review]
i get this wasnt your idea but something i brought up in the initial PR is that i feel like fibers and fingerprints are kind of arbitrarily separated and i dont see why they cant just be the same data, i.e. theyre just all 'traces of evidence' and the source doesn't matter

File: Content.Server/Forensics/Systems/ForensicPadSystem.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                    _popupSystem.PopupEntity(Loc.GetString("forensic-pad-start-scan-user", ("target", args.Target.Value)), args.Target.Value, Filter.Entities(args.User));
+                    _popupSystem.PopupEntity(Loc.GetString("forensic-pad-start-scan-target", ("user", args.User)), args.Target.Value, Filter.Entities(args.Target.Value));
+                }
+                StartScan(args.User, args.Target.Value, component, fingerprint.Fingerprint);
+                return;
+            }
+
+            if (TryComp<FiberComponent>(args.Target, out var fiber))
+                StartScan(args.User, args.Target.Value, component, Loc.GetString(fiber.FiberDescription));
+        }
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicPadSystem.cs:80]: args never get handled.

File: Content.Server/Forensics/Systems/ForensicPadSystem.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            args.PushMarkup(Loc.GetString("forensic-pad-sample", ("sample", component.Sample)));
+        }
+
+        private void OnAfterInteract(EntityUid uid, ForensicPadComponent component, AfterInteractEvent args)
+        {
+            if (component.CancelToken != null)
+            {
+                component.CancelToken.Cancel();
+                component.CancelToken = null;
+            }
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicPadSystem.cs:47]: In hindsight I don't like these and think it should just return if the canceltoken isn't null.

File: Content.Server/Forensics/Systems/ForensicPadSystem.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        private void OnTargetPadSuccessful(TargetPadSuccessfulEvent ev)
+        {
+            ev.Component.CancelToken = null;
+
+            PressSample(ev.Component.Owner, ev.Sample, ev.Component);
+        }
+        private void OnPadCancelled(PadCancelledEvent ev)
+        {
+            if (ev.Component == null)
+                return;
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicPadSystem.cs:114]: Would this component ever be null?

File: Content.Server/Forensics/Systems/ForensicPadSystem.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        public void PressSample(EntityUid uid, string sample, ForensicPadComponent? component = null)
+        {
+            if (!Resolve(uid, ref component))
+                return;
+
+            component.Sample = sample;
+            component.Used = true;
+        }
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicPadSystem.cs:103]: comment

File: Content.Server/Forensics/Systems/ForensicPadSystem.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+using Content.Shared.Examine;
+using Content.Shared.Interaction;
+using Content.Shared.Inventory;
+using Content.Server.DoAfter;
+using Content.Server.Popups;
+using Robust.Shared.Player;
+
+namespace Content.Server.Forensics
+{
+    public sealed class ForensicPadSystem : EntitySystem
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicPadSystem.cs:11]: comment, e.g. used to transfer fingerprints from target entities onto forensic pads.

File: Content.Server/Forensics/Systems/ForensicScannerSystem.cs [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            SubscribeLocalEvent<TargetScanSuccessfulEvent>(OnTargetScanSuccessful);
+            SubscribeLocalEvent<ScanCancelledEvent>(OnScanCancelled);
+        }
+
+        private void OnScanCancelled(ScanCancelledEvent ev)
+        {
+            if (ev.Component == null)
+                return;
+            ev.Component.CancelToken = null;
+        }
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicScannerSystem.cs:40]: As above.

File: Content.Server/Forensics/Systems/ForensicScannerSystem.cs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+            OpenUserInterface(ev.User, ev.Component);
+        }
+
+        private void OnAfterInteract(EntityUid uid, ForensicScannerComponent component, AfterInteractEvent args)
+        {
+            if (component.CancelToken != null)
+            {
+                component.CancelToken.Cancel();
+                component.CancelToken = null;
+            }
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicScannerSystem.cs:60]: As above.

File: Content.Server/Forensics/Systems/ForensicScannerSystem.cs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+            if (args.Handled || !args.CanReach)
+                return;
+
+            if (!TryComp<ForensicPadComponent>(args.Used, out var pad))
+                return;
+
+            foreach (var fiber in component.Fibers)
+            {
+                if (fiber == pad.Sample)
+                {
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicScannerSystem.cs:88]: This is kinda eh, how many fibers would a scanner expect? If it's more than like 10 it should be a hashset.

File: Content.Server/Forensics/Systems/ForensicScannerSystem.cs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+                    SoundSystem.Play(Filter.Pvs(uid), "/Audio/Machines/Nuke/angry_beep.ogg", uid);
```
Comment (metalgearsloth) [inline Content.Server/Forensics/Systems/ForensicScannerSystem.cs:89]: pass in entitymanager to filters until we come up with something less bad.

File: Content.Shared/Item/UserInteractedWithItemEvent.cs [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+
+/// <summary>
+/// Raised on the user after they do any sort of interaction with an item,
+/// useful for when you want a component on the user to do something to the item.
+/// E.g. forensics, disease, etc.
+/// </summary>
+public sealed class UserInteractedWithItemEvent : EntityEventArgs
+{
+    public EntityUid User;
+    public EntityUid Item;
```
Comment (metalgearsloth) [inline Content.Shared/Item/UserInteractedWithItemEvent.cs:11]: Might as well make it a ref event given it's all readonly value types anyway.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
A crumb of comments sire.

File: Content.Server/Forensics/Systems/ForensicScannerSystem.cs (hunk 12 above)
Comment (ike709) [reply Content.Server/Forensics/Systems/ForensicScannerSystem.cs:88]: Less than 10.

File: Content.Shared/Item/UserInteractedWithItemEvent.cs (hunk 14 above)
Comment (ike709) [reply Content.Shared/Item/UserInteractedWithItemEvent.cs:11]: I don't know what that means.

Comment (mirrorcult) [reply Content.Shared/Item/UserInteractedWithItemEvent.cs:11]: don't really think this is necessary

**Comment by ike709** [comment]
CI failures don't make sense, it builds & runs fine locally.

**Comment by EmoGarbage404** [comment]
> CI failures don't make sense, it builds & runs fine locally.

Its says you're passing strings into the filter param of some of the audio stuff. It doesn't even build.

**Comment by ike709** [comment]
> Its says you're passing strings into the filter param of some of the audio stuff. It doesn't even build.

Works fine locally, guess I need to merge master.

**Comment by themias** [comment]
I had this issue. There was a method signature update in the Robust Toolbox submodule, so make sure you're up to date with that.

**Comment by ike709** [comment]
There we go.

**Comment by moonheart08** [comment]
I'm so sorry ike.

**Comment by ike709** [comment]
One of these days it'll get reviewed again or merged.

**Review by moonheart08 — APPROVED** [review]

**Pull request merged**